okhttp/src/main/java/okhttp3/internal/connection/ExchangeFinder.kt

```diff
@@ -0,0 +1,310 @@
+    val readTimeout = chain.readTimeoutMillis()
+    val writeTimeout = chain.writeTimeoutMillis()
+    val pingIntervalMillis = client.pingIntervalMillis()
+    val connectionRetryEnabled = client.retryOnConnectionFailure()
+
+    try {
+      val resultConnection = findHealthyConnection(
+          connectTimeout = connectTimeout,
+          readTimeout = readTimeout,
+          writeTimeout = writeTimeout,
+          pingIntervalMillis = pingIntervalMillis,
+          connectionRetryEnabled = connectionRetryEnabled,
+          doExtensiveHealthChecks = doExtensiveHealthChecks
+      )
+      return resultConnection.newCodec(client, chain)
+    } catch (e: RouteException) {
+      trackFailure()
+      throw e
+    } catch (e: IOException) {
+      trackFailure()
+      throw RouteException(e)
+    }
+  }
+
+  /**
+   * Finds a connection and returns it if it is healthy. If it is unhealthy the process is repeated
+   * until a healthy connection is found.
+   */
+  @Throws(IOException::class)
+  private fun findHealthyConnection(
+    connectTimeout: Int,
+    readTimeout: Int,
+    writeTimeout: Int,
+    pingIntervalMillis: Int,
+    connectionRetryEnabled: Boolean,
+    doExtensiveHealthChecks: Boolean
+  ): RealConnection {
+    while (true) {
+      val candidate = findConnection(
+          connectTimeout = connectTimeout,
+          readTimeout = readTimeout,
+          writeTimeout = writeTimeout,
+          pingIntervalMillis = pingIntervalMillis,
+          connectionRetryEnabled = connectionRetryEnabled
+      )
+
+      // If this is a brand new connection, we can skip the extensive health checks.
+      synchronized(connectionPool) {
+        if (candidate.successCount == 0) {
+          return candidate
+        }
+      }
+
+      // Do a (potentially slow) check to confirm that the pooled connection is still good. If it
+      // isn't, take it out of the pool and start again.
+      if (!candidate.isHealthy(doExtensiveHealthChecks)) {
+        candidate.noNewExchanges()
+        continue
+      }
+
+      return candidate
+    }
+  }
+
+  /**
+   * Returns a connection to host a new stream. This prefers the existing connection if it exists,
+   * then the pool, finally building a new connection.
+   */
+  @Throws(IOException::class)
+  private fun findConnection(
+    connectTimeout: Int,
+    readTimeout: Int,
+    writeTimeout: Int,
+    pingIntervalMillis: Int,
+    connectionRetryEnabled: Boolean
+  ): RealConnection {
+    var foundPooledConnection = false
+    var result: RealConnection? = null
+    var selectedRoute: Route? = null
+    var releasedConnection: RealConnection?
+    val toClose: Socket?
+    synchronized(connectionPool) {
+      if (transmitter.isCanceled) throw IOException("Canceled")
+      hasStreamFailure = false // This is a fresh attempt.
+
+      val previousRoute = if (retryCurrentRoute()) {
+        // Attempt to use an already-allocated connection. We need to be careful here because our
```

Comment:
Rest of the comment is missing and should be in line 163. 

See: https://github.com/square/okhttp/pull/5023/files#diff-a2fc2ec01d45df9d9a8a6621cbfb4032L147

Reply:
Yeah Jesse you probably want
```
val previousRoute = if (retryCurrentRoute())  transmitter.connection!!.route() else null

// Attempt to use an already-allocated connection. We need to be careful here because our
// already-allocated connection may have been restricted from creating new exchanges.
releasedConnection = transmitter.connection
toClose = if (transmitter.connection?.noNewExchanges ?: false) {
      transmitter.releaseConnectionNoEvents()
} else {
      null
}
```

Reply:
Nice catch. Thanks!